Tests/SwiftAlgorithmsTests/TestUtilities.swift

```diff
@@ -66,3 +66,192 @@ func XCTAssertEqualSequences<S1: Sequence, S2: Sequence>(
 }
 
 func XCTAssertLazy<S: LazySequenceProtocol>(_: S) {}
+
+/// Tests that all index traversal methods behave as expected.
+///
+/// Verifies the correctness of the implementations of `startIndex`, `endIndex`,
+/// `indices`, `count`, `isEmpty`, `index(before:)`, `index(after:)`,
+/// `index(_:offsetBy:)`, `index(_:offsetBy:limitedBy:)`, and
+/// `distance(from:to:)` by calling them with just about all possible input
+/// combinations. When provided, the `indices` function is used to to test the
+/// collection methods against.
+///
+/// - Parameters:
+///   - collections: The collections to be validated.
+///   - indices: A closure that returns the expected indices of the given
+///     collection, including its `endIndex`, in ascending order. Only use this
+///     parameter if you are able to compute the indices of the collection
+///     independently of the `Collection` conformance, e.g. by using the
+///     contents of the collection directly.
+///
+/// - Complexity: O(*n*^3) for each collection, where *n* is the length of the
+///   collection.
+func validateIndexTraversals<C>(
+  _ collections: C...,
+  indices: ((C) -> [C.Index])? = nil,
+  file: StaticString = #file, line: UInt = #line
+) where C: BidirectionalCollection {
+  for c in collections {
+    let indices = indices?(c) ?? (c.indices + [c.endIndex])
+    let count = indices.count - 1
+    
+    XCTAssertEqual(
+      c.count, count,
+      "Count mismatch",
+      file: file, line: line)
+    XCTAssertEqual(
+      c.isEmpty, count == 0,
+      "Emptiness mismatch",
+      file: file, line: line)
+    XCTAssertEqual(
+      c.startIndex, indices.first,
+      "`startIndex` does not equal the first index",
+      file: file, line: line)
+    XCTAssertEqual(
+      c.endIndex, indices.last,
+      "`endIndex` does not equal the last index",
+      file: file, line: line)
+    
+    // `index(after:)`
+    do {
+      var index = c.startIndex
+      
+      for (offset, expected) in indices.enumerated().dropFirst() {
+        c.formIndex(after: &index)
+        XCTAssertEqual(
+          index, expected,
+          """
+          `startIndex` incremented \(offset) times does not equal index at \
+          offset \(offset)
+          """,
+          file: file, line: line)
+      }
+    }
+    
+    // `index(before:)`
+    do {
+      var index = c.endIndex
+
+      for (offset, expected) in indices.enumerated().dropLast().reversed() {
+        c.formIndex(before: &index)
+        XCTAssertEqual(
+          index, expected,
+          """
+          `endIndex` decremented \(count - offset) times does not equal index \
+          at offset \(offset)
+          """,
+          file: file, line: line)
+      }
+    }
+    
+    // `indices`
+    for (offset, index) in c.indices.enumerated() {
+      XCTAssertEqual(
+        index, indices[offset],
+        "Index mismatch at offset \(offset) in `indices`",
+        file: file, line: line)
+    }
```

Comment:
I think we need a check that `c.count` equals `c.indices.count` here.

Reply:
Nice find!